File: ios/RecipeClipperTests/Fakes/FakePantryRepository.swift
import Combine
import Foundation
@testable import RecipeClipper

/// A fake that models the pantry rather than only recording calls (Android's
/// FakePantryRepository): `items` is real state, a new item gets its aisle from the aisle table
/// as the real repository does, and snapshots restore.
final class FakePantryRepository: PantryRepository {
    let items: CurrentValueSubject<[PantryItem], Never>
    private var nextId: Int64

    init(_ initial: [PantryItem] = []) {
        items = CurrentValueSubject(initial)
        nextId = (initial.map(\.id).max() ?? 0) + 1
    }

    func observeItems() -> AnyPublisher<[PantryItem], Never> { items.eraseToAnyPublisher() }

    func items() async -> [PantryItem] { items.value }

    func add(_ item: NewPantryItem) async {
        let name = item.name.kTrimmed
        guard !name.isEmpty else { return }
        let quantity = item.quantity?.kTrimmed
        items.value.append(PantryItem(
            id: nextId, name: name, quantity: quantity?.isEmpty == false ? quantity : nil, language: item.language,
            aisle: item.aisle ?? Aisles.of(name, words: LanguageWords.forTag(item.language)),
            inStock: true, alwaysHave: false, purchasedDay: item.purchasedDay, expiresDay: nil
        ))
        nextId += 1
    }

    func setInStock(_ ids: [Int64], inStock: Bool) async {
        items.value = items.value.map { var i = $0; if ids.contains(i.id) { i.inStock = inStock }; return i }
    }

    func restock(_ ids: [Int64], day: Int64) async {
        items.value = items.value.map { var i = $0; if ids.contains(i.id) { i.inStock = true; i.purchasedDay = day }; return i }
    }

    func edit(_ id: Int64, _ edit: PantryEdit) async {
        let name = edit.name.kTrimmed
        guard !name.isEmpty else { return }
        let quantity = edit.quantity?.kTrimmed
        items.value = items.value.map {
            var i = $0
            if i.id == id {
                i.name = name
                i.quantity = quantity?.isEmpty == false ? quantity : nil
                i.alwaysHave = edit.alwaysHave
                i.expiresDay = edit.expiresDay
            }
            return i
        }
    }

    func snapshot(_ ids: [Int64]) async -> PantrySnapshot {
        PantrySnapshot(items: items.value.filter { ids.contains($0.id) }.map(Self.record))
    }

    func delete(_ id: Int64) async -> PantrySnapshot? {
        guard let gone = items.value.first(where: { $0.id == id }) else { return nil }
        items.value.removeAll { $0.id == id }
        return PantrySnapshot(items: [Self.record(gone)])
    }

    func restore(_ snapshot: PantrySnapshot) async {
        let back = snapshot.items.map(\.domain)
        items.value = (items.value.filter { item in !back.contains { $0.id == item.id } } + back).sorted { $0.id < $1.id }
    }

    private static func record(_ i: PantryItem) -> PantryItemRecord {
        PantryItemRecord(
            id: i.id, name: i.name, quantity: i.quantity, language: i.language, aisle: i.aisle.key, inStock: i.inStock,
            alwaysHave: i.alwaysHave, purchasedDay: i.purchasedDay, expiresDay: i.expiresDay, updatedAt: 0
        )
    }
}

File: app/src/test/java/com/example/recipeclipper/fake/FakePantryRepository.kt
package com.example.recipeclipper.fake

import com.example.recipeclipper.data.PantryRepository
import com.example.recipeclipper.data.local.entity.PantryItemEntity
import com.example.recipeclipper.data.model.Aisle
import com.example.recipeclipper.data.model.Aisles
import com.example.recipeclipper.data.model.LanguageWords
import com.example.recipeclipper.data.model.NewPantryItem
import com.example.recipeclipper.data.model.PantryEdit
import com.example.recipeclipper.data.model.PantryItem
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow

/**
 * A fake that models the pantry rather than only recording calls: [items] is real state, a new
 * item gets its aisle from the aisle table as the real repository does, and snapshots restore.
 */
class FakePantryRepository(initial: List<PantryItem> = emptyList()) : PantryRepository {

    val items = MutableStateFlow(initial)

    private var nextId = (initial.maxOfOrNull { it.id } ?: 0L) + 1

    override fun observeItems(): Flow<List<PantryItem>> = items

    override suspend fun items(): List<PantryItem> = items.value

    override suspend fun add(item: NewPantryItem) {
        val name = item.name.trim().takeIf { it.isNotEmpty() } ?: return
        items.value = items.value + PantryItem(
            id = nextId++,
            name = name,
            quantity = item.quantity?.trim()?.takeIf { it.isNotEmpty() },
            language = item.language,
            aisle = item.aisle ?: Aisles.of(name, LanguageWords.forTag(item.language)),
            inStock = true,
            alwaysHave = false,
            purchasedDay = item.purchasedDay,
            expiresDay = null
        )
    }

    override suspend fun setInStock(ids: List<Long>, inStock: Boolean) {
        items.value = items.value.map { if (it.id in ids) it.copy(inStock = inStock) else it }
    }

    override suspend fun restock(ids: List<Long>, day: Long) {
        items.value = items.value.map { if (it.id in ids) it.copy(inStock = true, purchasedDay = day) else it }
    }

    override suspend fun edit(id: Long, edit: PantryEdit) {
        val name = edit.name.trim().takeIf { it.isNotEmpty() } ?: return
        items.value = items.value.map {
            if (it.id == id) it.copy(
                name = name,
                quantity = edit.quantity?.trim()?.takeIf { q -> q.isNotEmpty() },
                alwaysHave = edit.alwaysHave,
                expiresDay = edit.expiresDay
            ) else it
        }
    }

    override suspend fun snapshot(ids: List<Long>): PantryRepository.Snapshot =
        PantryRepository.Snapshot(items.value.filter { it.id in ids }.map { it.toEntity() })

    override suspend fun delete(id: Long): PantryRepository.Snapshot? {
        val gone = items.value.firstOrNull { it.id == id } ?: return null
        items.value = items.value - gone
        return PantryRepository.Snapshot(listOf(gone.toEntity()))
    }

    override suspend fun restore(snapshot: PantryRepository.Snapshot) {
        val back = snapshot.entities.map { it.toDomain() }
        items.value = (items.value.filterNot { item -> back.any { it.id == item.id } } + back).sortedBy { it.id }
    }

    private fun PantryItem.toEntity() = PantryItemEntity(
        id = id, name = name, quantity = quantity, language = language, aisle = aisle.key, inStock = inStock,
        alwaysHave = alwaysHave, purchasedDay = purchasedDay, expiresDay = expiresDay, updatedAt = 0
    )

    private fun PantryItemEntity.toDomain() = PantryItem(
        id = id, name = name, quantity = quantity, language = language, aisle = Aisle.fromKey(aisle), inStock = inStock,
        alwaysHave = alwaysHave, purchasedDay = purchasedDay, expiresDay = expiresDay
    )
}
